File: packages/typescript/src/tracer.ts
/**
 * Execution Tracer — structured tracing for pipeline execution.
 */

import { randomUUID } from "node:crypto";
import type { AgentRole } from "./types.js";

export interface Span {
  id: string;
  traceId: string;
  parentId?: string;
  name: string;
  role?: AgentRole;
  action?: string;
  startTime: number;
  endTime?: number;
  durationMs?: number;
  status: "running" | "completed" | "failed";
  metadata: Record<string, unknown>;
}

export interface Trace {
  id: string;
  taskId: string;
  pipelineName: string;
  startTime: number;
  endTime?: number;
  durationMs?: number;
  spans: Span[];
  metadata: Record<string, unknown>;
}

export class Tracer {
  private traces: Map<string, Trace> = new Map();

  /**
   * Start a new trace for a pipeline execution.
   */
  startTrace(taskId: string, pipelineName: string): Trace {
    const trace: Trace = {
      id: randomUUID(),
      taskId,
      pipelineName,
      startTime: Date.now(),
      spans: [],
      metadata: {},
    };
    this.traces.set(trace.id, trace);
    return trace;
  }

  /**
   * End a trace.
   */
  endTrace(traceId: string): Trace | undefined {
    const trace = this.traces.get(traceId);
    if (trace) {
      trace.endTime = Date.now();
      trace.durationMs = trace.endTime - trace.startTime;
    }
    return trace;
  }

  /**
   * Start a new span within a trace.
   */
  startSpan(
    traceId: string,
    name: string,
    options: {
      role?: AgentRole;
      action?: string;
      parentId?: string;
    } = {}
  ): Span {
    const span: Span = {
      id: randomUUID(),
      traceId,
      parentId: options.parentId,
      name,
      role: options.role,
      action: options.action,
      startTime: Date.now(),
      status: "running",
      metadata: {},
    };

    const trace = this.traces.get(traceId);
    if (trace) {
      trace.spans.push(span);
    }

    return span;
  }

  /**
   * End a span.
   */
  endSpan(
    span: Span,
    status: "completed" | "failed" = "completed"
  ): void {
    span.endTime = Date.now();
    span.durationMs = span.endTime - span.startTime;
    span.status = status;
  }

  /**
   * Get a trace by ID.
   */
  getTrace(traceId: string): Trace | undefined {
    return this.traces.get(traceId);
  }

  /**
   * Get all traces.
   */
  getAllTraces(): Trace[] {
    return Array.from(this.traces.values());
  }

  /**
   * Export a trace as JSON.
   */
  exportTraceJson(traceId: string): string | null {
    const trace = this.traces.get(traceId);
    if (!trace) return null;
    return JSON.stringify(trace, null, 2);
  }

  /**
   * Print a human-readable trace summary to console.
   */
  printTraceSummary(traceId: string): void {
    const trace = this.traces.get(traceId);
    if (!trace) {
      console.log("Trace not found:", traceId);
      return;
    }

    console.log(`\n📊 Trace: ${trace.pipelineName} (${trace.taskId})`);
    console.log(`   ID: ${trace.id}`);
    console.log(
      `   Duration: ${trace.durationMs ? `${trace.durationMs}ms` : "running"}`
    );
    console.log(`   Spans: ${trace.spans.length}\n`);

    for (const span of trace.spans) {
      const status =
        span.status === "completed"
          ? "✅"
          : span.status === "failed"
            ? "❌"
            : "⏳";
      const duration = span.durationMs ? `${span.durationMs}ms` : "...";
      const role = span.role ? ` [${span.role}]` : "";
      console.log(`   ${status} ${span.name}${role} — ${duration}`);
    }
    console.log();
  }

  /**
   * Clear all traces.
   */
  clear(): void {
    this.traces.clear();
  }
}

File: packages/python/src/tcs_bpi_ai_team/tracer.py
"""Execution Tracer — structured tracing for pipeline execution."""

from __future__ import annotations

import json
import time
import uuid
from dataclasses import dataclass, field

from tcs_bpi_ai_team.types import AgentRole


@dataclass
class Span:
    """A single execution span within a trace."""

    id: str
    trace_id: str
    name: str
    start_time: float
    parent_id: str | None = None
    role: AgentRole | None = None
    action: str | None = None
    end_time: float | None = None
    duration_ms: float | None = None
    status: str = "running"  # "running" | "completed" | "failed"
    metadata: dict[str, object] = field(default_factory=dict)


@dataclass
class Trace:
    """A full execution trace for a pipeline run."""

    id: str
    task_id: str
    pipeline_name: str
    start_time: float
    end_time: float | None = None
    duration_ms: float | None = None
    spans: list[Span] = field(default_factory=list)
    metadata: dict[str, object] = field(default_factory=dict)


class Tracer:
    """Execution tracer for pipeline operations."""

    def __init__(self) -> None:
        self._traces: dict[str, Trace] = {}

    def start_trace(self, task_id: str, pipeline_name: str) -> Trace:
        """Start a new trace for a pipeline execution."""
        trace = Trace(
            id=str(uuid.uuid4()),
            task_id=task_id,
            pipeline_name=pipeline_name,
            start_time=time.time() * 1000,
        )
        self._traces[trace.id] = trace
        return trace

    def end_trace(self, trace_id: str) -> Trace | None:
        """End a trace."""
        trace = self._traces.get(trace_id)
        if trace:
            trace.end_time = time.time() * 1000
            trace.duration_ms = trace.end_time - trace.start_time
        return trace

    def start_span(
        self,
        trace_id: str,
        name: str,
        *,
        role: AgentRole | None = None,
        action: str | None = None,
        parent_id: str | None = None,
    ) -> Span:
        """Start a new span within a trace."""
        span = Span(
            id=str(uuid.uuid4()),
            trace_id=trace_id,
            name=name,
            start_time=time.time() * 1000,
            role=role,
            action=action,
            parent_id=parent_id,
        )
        trace = self._traces.get(trace_id)
        if trace:
            trace.spans.append(span)
        return span

    def end_span(self, span: Span, status: str = "completed") -> None:
        """End a span."""
        span.end_time = time.time() * 1000
        span.duration_ms = span.end_time - span.start_time
        span.status = status

    def get_trace(self, trace_id: str) -> Trace | None:
        """Get a trace by ID."""
        return self._traces.get(trace_id)

    def get_all_traces(self) -> list[Trace]:
        """Get all traces."""
        return list(self._traces.values())

    def export_trace_json(self, trace_id: str) -> str | None:
        """Export a trace as JSON."""
        trace = self._traces.get(trace_id)
        if trace is None:
            return None

        def _serialize(obj: object) -> object:
            if isinstance(obj, Span | Trace):
                return {k: v for k, v in obj.__dict__.items()}
            return str(obj)

        return json.dumps(trace.__dict__, default=_serialize, indent=2)

    def print_trace_summary(self, trace_id: str) -> None:
        """Print a human-readable trace summary."""
        trace = self._traces.get(trace_id)
        if trace is None:
            print(f"Trace not found: {trace_id}")
            return

        duration = f"{trace.duration_ms:.0f}ms" if trace.duration_ms else "running"
        print(f"\n📊 Trace: {trace.pipeline_name} ({trace.task_id})")
        print(f"   ID: {trace.id}")
        print(f"   Duration: {duration}")
        print(f"   Spans: {len(trace.spans)}\n")

        for span in trace.spans:
            icon = {"completed": "✅", "failed": "❌"}.get(span.status, "⏳")
            dur = f"{span.duration_ms:.0f}ms" if span.duration_ms else "..."
            role_str = f" [{span.role}]" if span.role else ""
            print(f"   {icon} {span.name}{role_str} — {dur}")

        print()

    def clear(self) -> None:
        """Clear all traces."""
        self._traces.clear()
